File: frontend/lib/api/stateWorklist.ts
import apiClient from './client'

// --- State Assessment Types ---

export interface StateAssessment {
  id: string
  client_id: string
  engagement_id?: string
  analysis_id?: string
  organization_id: string
  state: string
  nexus_status: 'unknown' | 'needs_data' | 'no_nexus' | 'approaching' | 'has_nexus' | 'excluded'
  nexus_type?: 'economic' | 'physical' | 'both'
  nexus_reasons: string[]
  first_exposure_date?: string
  threshold_percentage?: number
  total_sales?: number
  estimated_liability?: number
  confidence_level?: 'high' | 'medium' | 'low'
  notes?: string
  assessed_at?: string
  assessed_by?: string
  created_at: string
  updated_at: string
  current_action?: StateAction
}

export interface StateAssessmentCreate {
  client_id: string
  state: string
  engagement_id?: string
  analysis_id?: string
  nexus_status?: string
  nexus_type?: string
  nexus_reasons?: string[]
  first_exposure_date?: string
  threshold_percentage?: number
  total_sales?: number
  estimated_liability?: number
  confidence_level?: string
  notes?: string
}

export interface StateAssessmentUpdate {
  nexus_status?: string
  nexus_type?: string
  nexus_reasons?: string[]
  first_exposure_date?: string
  threshold_percentage?: number
  total_sales?: number
  estimated_liability?: number
  confidence_level?: string
  notes?: string
}

// --- State Action Types ---

export type ActionType = 'register' | 'vda' | 'file_back_returns' | 'marketplace_exception' | 'monitor' | 'no_action'
export type ActionStatus = 'not_started' | 'in_progress' | 'blocked' | 'complete'

export interface StateAction {
  id: string
  state_assessment_id: string
  organization_id: string
  action_type: ActionType
  action_status: ActionStatus
  blocked_reason?: string
  blocked_since?: string
  assigned_to?: string
  target_date?: string
  strategy_notes?: string
  registration_effective_date?: string
  vda_submission_date?: string
  vda_approval_date?: string
  completed_at?: string
  created_at: string
  updated_at: string
  tasks: StateActionTask[]
}

export interface StateActionCreate {
  state_assessment_id: string
  action_type: ActionType
  action_status?: ActionStatus
  assigned_to?: string
  target_date?: string
  strategy_notes?: string
}

export interface StateActionUpdate {
  action_type?: ActionType
  action_status?: ActionStatus
  blocked_reason?: string
  blocked_since?: string
  assigned_to?: string
  target_date?: string
  strategy_notes?: string
  registration_effective_date?: string
  vda_submission_date?: string
  vda_approval_date?: string
}

// --- State Action Task Types ---

export type TaskStatus = 'pending' | 'in_progress' | 'complete' | 'skipped'

export interface StateActionTask {
  id: string
  state_action_id: string
  title: string
  description?: string
  status: TaskStatus
  assigned_to?: string
  due_date?: string
  completed_at?: string
  sort_order: number
  created_at: string
  updated_at: string
}

export interface StateActionTaskCreate {
  state_action_id: string
  title: string
  description?: string
  status?: TaskStatus
  assigned_to?: string
  due_date?: string
  sort_order?: number
}

export interface StateActionTaskUpdate {
  title?: string
  description?: string
  status?: TaskStatus
  assigned_to?: string
  due_date?: string
  sort_order?: number
}

// --- Worklist Types ---

export interface StateWorklistItem {
  id: string
  state: string
  nexus_status: string
  nexus_type?: string
  nexus_reasons: string[]
  estimated_liability?: number
  threshold_percentage?: number
  first_exposure_date?: string
  action_id?: string
  action_type?: ActionType
  action_status?: ActionStatus
  target_date?: string
  blocked_reason?: string
  tasks_total: number
  tasks_complete: number
}

export interface StateWorklistSummary {
  total_states: number
  states_with_nexus: number
  states_approaching: number
  states_no_nexus: number
  states_unknown: number
  actions_not_started: number
  actions_in_progress: number
  actions_blocked: number
  actions_complete: number
  total_estimated_liability: number
}

export interface StateWorklistResponse {
  items: StateWorklistItem[]
  summary: StateWorklistSummary
}

// --- API Functions ---

// State Worklist
export async function getStateWorklist(
  clientId: string,
  filters?: { nexus_status?: string; action_status?: string }
): Promise<StateWorklistResponse> {
  const params = new URLSearchParams()
  if (filters?.nexus_status) params.append('nexus_status', filters.nexus_status)
  if (filters?.action_status) params.append('action_status', filters.action_status)
  const queryString = params.toString()
  const url = `/api/v1/clients/${clientId}/states${queryString ? `?${queryString}` : ''}`
  const response = await apiClient.get(url)
  return response.data
}

// State Assessments
export async function createStateAssessment(
  clientId: string,
  data: StateAssessmentCreate
): Promise<StateAssessment> {
  const response = await apiClient.post(`/api/v1/clients/${clientId}/states`, data)
  return response.data
}

export async function getStateAssessment(
  clientId: string,
  state: string
): Promise<StateAssessment> {
  const response = await apiClient.get(`/api/v1/clients/${clientId}/states/${state}`)
  return response.data
}

export async function updateStateAssessment(
  clientId: string,
  state: string,
  data: StateAssessmentUpdate
): Promise<StateAssessment> {
  const response = await apiClient.patch(`/api/v1/clients/${clientId}/states/${state}`, data)
  return response.data
}

export async function deleteStateAssessment(
  clientId: string,
  state: string
): Promise<void> {
  await apiClient.delete(`/api/v1/clients/${clientId}/states/${state}`)
}

// State Actions
export async function createStateAction(
  clientId: string,
  state: string,
  data: StateActionCreate,
  autoCreateTasks: boolean = true
): Promise<StateAction> {
  const response = await apiClient.post(
    `/api/v1/clients/${clientId}/states/${state}/actions?auto_create_tasks=${autoCreateTasks}`,
    data
  )
  return response.data
}

export async function updateStateAction(
  clientId: string,
  state: string,
  actionId: string,
  data: StateActionUpdate
): Promise<StateAction> {
  const response = await apiClient.patch(
    `/api/v1/clients/${clientId}/states/${state}/actions/${actionId}`,
    data
  )
  return response.data
}

export async function deleteStateAction(
  clientId: string,
  state: string,
  actionId: string
): Promise<void> {
  await apiClient.delete(`/api/v1/clients/${clientId}/states/${state}/actions/${actionId}`)
}

// State Action Tasks
export async function createActionTask(
  clientId: string,
  state: string,
  actionId: string,
  data: StateActionTaskCreate
): Promise<StateActionTask> {
  const response = await apiClient.post(
    `/api/v1/clients/${clientId}/states/${state}/actions/${actionId}/tasks`,
    data
  )
  return response.data
}

export async function updateActionTask(
  clientId: string,
  state: string,
  actionId: string,
  taskId: string,
  data: StateActionTaskUpdate
): Promise<StateActionTask> {
  const response = await apiClient.patch(
    `/api/v1/clients/${clientId}/states/${state}/actions/${actionId}/tasks/${taskId}`,
    data
  )
  return response.data
}

export async function deleteActionTask(
  clientId: string,
  state: string,
  actionId: string,
  taskId: string
): Promise<void> {
  await apiClient.delete(
    `/api/v1/clients/${clientId}/states/${state}/actions/${actionId}/tasks/${taskId}`
  )
}

// Import from Analysis
export async function importStatesFromAnalysis(
  clientId: string,
  analysisId: string
): Promise<{ imported: number; skipped: number; total_states: number }> {
  const response = await apiClient.post(
    `/api/v1/clients/${clientId}/states/import-from-analysis?analysis_id=${analysisId}`
  )
  return response.data
}

// --- Display Helpers ---

export const NEXUS_STATUS_LABELS: Record<string, { label: string; color: string }> = {
  unknown: { label: 'Unknown', color: 'gray' },
  needs_data: { label: 'Needs Data', color: 'yellow' },
  no_nexus: { label: 'No Nexus', color: 'green' },
  approaching: { label: 'Approaching', color: 'orange' },
  has_nexus: { label: 'Has Nexus', color: 'red' },
  excluded: { label: 'Excluded', color: 'gray' },
}

export const ACTION_TYPE_LABELS: Record<ActionType, { label: string; description: string }> = {
  register: { label: 'Register', description: 'Register for sales tax collection' },
  vda: { label: 'VDA', description: 'Voluntary Disclosure Agreement' },
  file_back_returns: { label: 'File Back Returns', description: 'File delinquent returns' },
  marketplace_exception: { label: 'Marketplace', description: 'Marketplace facilitator exception' },
  monitor: { label: 'Monitor', description: 'Monitor threshold progress' },
  no_action: { label: 'No Action', description: 'No action required' },
}

export const ACTION_STATUS_LABELS: Record<ActionStatus, { label: string; color: string }> = {
  not_started: { label: 'Not Started', color: 'gray' },
  in_progress: { label: 'In Progress', color: 'blue' },
  blocked: { label: 'Blocked', color: 'red' },
  complete: { label: 'Complete', color: 'green' },
}

export const TASK_STATUS_LABELS: Record<TaskStatus, { label: string; color: string }> = {
  pending: { label: 'Pending', color: 'gray' },
  in_progress: { label: 'In Progress', color: 'blue' },
  complete: { label: 'Complete', color: 'green' },
  skipped: { label: 'Skipped', color: 'gray' },
}

File: backend/app/schemas/state_worklist.py
"""Schemas for state worklist and action tracking."""

from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime, date
from uuid import UUID


# --- State Assessment Schemas ---

class StateAssessmentBase(BaseModel):
    """Base fields for state assessment."""
    state: str = Field(..., min_length=2, max_length=2)
    nexus_status: str = "unknown"  # unknown, needs_data, no_nexus, approaching, has_nexus, excluded
    nexus_type: Optional[str] = None  # economic, physical, both
    nexus_reasons: List[str] = []
    first_exposure_date: Optional[date] = None
    threshold_percentage: Optional[float] = None
    total_sales: Optional[float] = None
    estimated_liability: Optional[float] = None
    confidence_level: Optional[str] = None  # high, medium, low
    notes: Optional[str] = None


class StateAssessmentCreate(StateAssessmentBase):
    """Create a state assessment."""
    client_id: UUID
    engagement_id: Optional[UUID] = None
    analysis_id: Optional[UUID] = None


class StateAssessmentUpdate(BaseModel):
    """Update a state assessment."""
    nexus_status: Optional[str] = None
    nexus_type: Optional[str] = None
    nexus_reasons: Optional[List[str]] = None
    first_exposure_date: Optional[date] = None
    threshold_percentage: Optional[float] = None
    total_sales: Optional[float] = None
    estimated_liability: Optional[float] = None
    confidence_level: Optional[str] = None
    notes: Optional[str] = None


class StateAssessmentResponse(StateAssessmentBase):
    """State assessment response with metadata."""
    id: UUID
    client_id: UUID
    engagement_id: Optional[UUID] = None
    analysis_id: Optional[UUID] = None
    organization_id: UUID
    assessed_at: Optional[datetime] = None
    assessed_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    # Include current action if exists
    current_action: Optional["StateActionResponse"] = None

    class Config:
        from_attributes = True


# --- State Action Schemas ---

class StateActionBase(BaseModel):
    """Base fields for state action."""
    action_type: str  # register, vda, file_back_returns, marketplace_exception, monitor, no_action
    action_status: str = "not_started"  # not_started, in_progress, blocked, complete
    blocked_reason: Optional[str] = None
    blocked_since: Optional[date] = None
    assigned_to: Optional[UUID] = None
    target_date: Optional[date] = None
    strategy_notes: Optional[str] = None
    registration_effective_date: Optional[date] = None
    vda_submission_date: Optional[date] = None
    vda_approval_date: Optional[date] = None


class StateActionCreate(StateActionBase):
    """Create a state action."""
    state_assessment_id: UUID


class StateActionUpdate(BaseModel):
    """Update a state action."""
    action_type: Optional[str] = None
    action_status: Optional[str] = None
    blocked_reason: Optional[str] = None
    blocked_since: Optional[date] = None
    assigned_to: Optional[UUID] = None
    target_date: Optional[date] = None
    strategy_notes: Optional[str] = None
    registration_effective_date: Optional[date] = None
    vda_submission_date: Optional[date] = None
    vda_approval_date: Optional[date] = None


class StateActionResponse(StateActionBase):
    """State action response with metadata."""
    id: UUID
    state_assessment_id: UUID
    organization_id: UUID
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    # Include tasks
    tasks: List["StateActionTaskResponse"] = []

    class Config:
        from_attributes = True


# --- State Action Task Schemas ---

class StateActionTaskBase(BaseModel):
    """Base fields for action task."""
    title: str
    description: Optional[str] = None
    status: str = "pending"  # pending, in_progress, complete, skipped
    assigned_to: Optional[UUID] = None
    due_date: Optional[date] = None
    sort_order: int = 0


class StateActionTaskCreate(StateActionTaskBase):
    """Create an action task."""
    state_action_id: UUID


class StateActionTaskUpdate(BaseModel):
    """Update an action task."""
    title: Optional[str] = None
    description: Optional[str] = None
    status: Optional[str] = None
    assigned_to: Optional[UUID] = None
    due_date: Optional[date] = None
    sort_order: Optional[int] = None


class StateActionTaskResponse(StateActionTaskBase):
    """Action task response."""
    id: UUID
    state_action_id: UUID
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# --- Worklist Summary Schemas ---

class StateWorklistItem(BaseModel):
    """Combined view for state worklist table."""
    id: UUID  # assessment id
    state: str
    nexus_status: str
    nexus_type: Optional[str] = None
    nexus_reasons: List[str] = []
    estimated_liability: Optional[float] = None
    threshold_percentage: Optional[float] = None
    first_exposure_date: Optional[date] = None

    # Current action info (if exists)
    action_id: Optional[UUID] = None
    action_type: Optional[str] = None
    action_status: Optional[str] = None
    target_date: Optional[date] = None
    blocked_reason: Optional[str] = None

    # Task progress
    tasks_total: int = 0
    tasks_complete: int = 0


class StateWorklistSummary(BaseModel):
    """Summary stats for the worklist."""
    total_states: int = 0
    states_with_nexus: int = 0
    states_approaching: int = 0
    states_no_nexus: int = 0
    states_unknown: int = 0

    actions_not_started: int = 0
    actions_in_progress: int = 0
    actions_blocked: int = 0
    actions_complete: int = 0

    total_estimated_liability: float = 0


class StateWorklistResponse(BaseModel):
    """Full worklist response."""
    items: List[StateWorklistItem]
    summary: StateWorklistSummary


# --- Action Templates ---

class ActionTaskTemplate(BaseModel):
    """Template for auto-creating tasks when an action is created."""
    title: str
    description: Optional[str] = None
    sort_order: int = 0


# Default task templates by action type
ACTION_TASK_TEMPLATES: dict[str, List[ActionTaskTemplate]] = {
    "register": [
        ActionTaskTemplate(title="Gather registration documents", sort_order=1),
        ActionTaskTemplate(title="Complete state registration application", sort_order=2),
        ActionTaskTemplate(title="Submit registration to state", sort_order=3),
        ActionTaskTemplate(title="Receive confirmation/permit number", sort_order=4),
        ActionTaskTemplate(title="Set up tax engine for state", sort_order=5),
    ],
    "vda": [
        ActionTaskTemplate(title="Review VDA program eligibility", sort_order=1),
        ActionTaskTemplate(title="Calculate lookback period liability", sort_order=2),
        ActionTaskTemplate(title="Prepare VDA application", sort_order=3),
        ActionTaskTemplate(title="Draft anonymous pre-clearance (if applicable)", sort_order=4),
        ActionTaskTemplate(title="Submit VDA application", sort_order=5),
        ActionTaskTemplate(title="Negotiate settlement terms", sort_order=6),
        ActionTaskTemplate(title="Execute VDA agreement", sort_order=7),
        ActionTaskTemplate(title="Submit payment", sort_order=8),
        ActionTaskTemplate(title="Register for prospective filing", sort_order=9),
    ],
    "file_back_returns": [
        ActionTaskTemplate(title="Determine filing periods owed", sort_order=1),
        ActionTaskTemplate(title="Calculate tax due per period", sort_order=2),
        ActionTaskTemplate(title="Prepare back returns", sort_order=3),
        ActionTaskTemplate(title="Submit returns to state", sort_order=4),
        ActionTaskTemplate(title="Submit payment with interest/penalties", sort_order=5),
    ],
    "marketplace_exception": [
        ActionTaskTemplate(title="Document marketplace facilitator status", sort_order=1),
        ActionTaskTemplate(title="Verify state marketplace facilitator law", sort_order=2),
        ActionTaskTemplate(title="Confirm no direct sales nexus", sort_order=3),
        ActionTaskTemplate(title="Document decision rationale", sort_order=4),
    ],
    "monitor": [
        ActionTaskTemplate(title="Set threshold monitoring alert", sort_order=1),
        ActionTaskTemplate(title="Document current sales levels", sort_order=2),
        ActionTaskTemplate(title="Schedule quarterly review", sort_order=3),
    ],
}


# Update forward references
StateAssessmentResponse.model_rebuild()
StateActionResponse.model_rebuild()
